Fix/normalized hypergraph laplacian

kaiser-dan opened the pull request:
# Summary
Rewrote `normalized_hypergraph_laplacian` implementation to match reference implementation. Fixes #647 .

# Description
Rewrote the matrix calculations in `normalized_hypergraph_laplacian` to match the implementation referenced here: https://github.com/xgi-org/xgi/issues/647#issuecomment-2612087221. Also adds a `weighted` Boolean parameter to allow for differentially weighting edges in the Laplacian calculation, drawn from each edge's "weight" attribute if available (default to 1 if unavailable).

# Concerns
I have only added barebone unit tests for the _effect_ of edge weights (only on their acceptable definition). I am unsure how or what the expected behavior would be for non-uniform edge weights.

# Other
~~Currently only supports strictly positive weights - a weight of zero would cause some issues with rank. That said, an edge weight of 0 could be interpreted as the subgraph with that edge removed and could be a nice way to handle zero weights in the future?~~ Deprecated with latest commit [573eaad](https://github.com/xgi-org/xgi/pull/648/commits/573eaad7c41708e9c0e640c613d3746588147014).

maximelucas commented on tests/linalg/test_matrix.py:
```diff
@@ -605,24 +607,47 @@ def test_normalized_hypergraph_laplacian():
-    assert np.all(np.diag(L2) == 0.5)
 
-    L3, d = xgi.normalized_hypergraph_laplacian(H, index=True)
+    evals = eigh(L2, eigvals_only=True)
```
if you don't need the eigenvalues, you can use `eigvalsh`, which outputs them in ascending order. I think it's supposed to be faster

maximelucas commented on xgi/linalg/laplacian_matrix.py:
```diff
@@ -221,16 +234,60 @@ def normalized_hypergraph_laplacian(H, sparse=True, index=False):
-    D = degree_matrix(H)
-    A, rowdict = clique_motif_matrix(H, sparse=sparse, index=True)
+    if weights is not None:
+        if weights is not list:
```
i don't think this works? better to use isinstance to check for type

kaiser-dan replied:
Will be obsolete with move to boolean indicator for weights and edge attributes

maximelucas commented on xgi/linalg/laplacian_matrix.py:
```diff
@@ -221,16 +234,60 @@ def normalized_hypergraph_laplacian(H, sparse=True, index=False):
-        Dinvsqrt = np.diag(np.power(D, -0.5))
+        Dv_invsqrt = np.diag(np.power(Dv, -0.5))
+
+        De_inv = np.diag(list(
+            map(
+                lambda x: 1/x,
+                np.sum(H_, axis=0)
+            )
+        ))
```
```suggestion
        De = np.sum(H_, axis=0)
        De_inv = np.diag(1 / De)
```
A bit more legible and hopefully a bit faster

maximelucas commented on xgi/linalg/laplacian_matrix.py:
```diff
@@ -221,16 +234,60 @@ def normalized_hypergraph_laplacian(H, sparse=True, index=False):
+
+    Dv = degree_matrix(H)
+    (
+        H_,
```
I was confusing H_ with a hypergraph, maybe replace by `incidence`  or other (even if they call it H in the paper

maximelucas commented on xgi/linalg/laplacian_matrix.py:
```diff
@@ -221,16 +234,60 @@ def normalized_hypergraph_laplacian(H, sparse=True, index=False):
+    # PERF: There is a faster way to do this calculation if unweighted.
+    # W can be ignored entirely if unweighted, but it adds a couple conditionals and complicates the code.
+    # It is untested, but I suspect the performance change is negligible.
+    L = (eye - (Dv_invsqrt @ H_ @ W @ De_inv @ np.transpose(H_) @ Dv_invsqrt))
```
does this np.transpose function work when H_ is a sparse array?

nwlandry replied:
The performance comment should be moved to an issue or solved with this PR.

kaiser-dan replied:
I will remove it from the next update and make an improvement issue if/when this PR is approved

maximelucas commented on tests/linalg/test_matrix.py:
```diff
@@ -605,24 +607,47 @@ def test_normalized_hypergraph_laplacian():
+    negative_evals = list(filter(lambda e: e < 0, evals))
+    assert (not negative_evals) or (np.allclose(negative_evals, 0))
 
+    L3, d = xgi.normalized_hypergraph_laplacian(H, index=True)
```
values or at least eigenvalues need to be tested in this case (only sparse case)

maximelucas commented on xgi/linalg/laplacian_matrix.py:
```diff
@@ -221,16 +234,60 @@ def normalized_hypergraph_laplacian(H, sparse=True, index=False):
-        Dinvsqrt = csr_array(diags(np.power(D, -0.5)))
+        Dv_invsqrt = csr_array(diags(np.power(Dv, -0.5)))
+
+        De_inv = diags(list(
+            map(
+                lambda x: 1/x,
+                np.sum(H_, axis=0)
+            )
+        ))
```
maybe try to simplify similarly to the non-sparse case. Also, is np.sum(H_, axis=0) still sparse?

maximelucas commented:
Thanks for the PR!

It looks good, I only left a few comments to check that everything is okay in the sparse case, and to simplify a few lines. 

About the weights, I'm thinking maybe we want the argument to be boolean instead: `weighted=False`. When it's True, we could just get the weights from the "weight" attribute of the hyperedges encoded in the hypergraph.
What do you think? Or do you see a case where one would want to use other weights?

That's how we have it for the adjacency matrix now.

kaiser-dan commented:
Okie I believe this is ready for re-review. I have addressed most of your comments @maximelucas  by directly implementing the suggestions, however, to clarify on the remaining, here are two observations:
- the `np.transpose` of a SciPy sparse array is still a sparse array
- `np.sum` of a sparse matrix is _not_ sparse, so in the middle of the `De` calculations a dense array is made, then cast back into a sparse diagonal matrix. This is introducing an additional memory allocation of $\Theta(m)$, where m is the number of hyperedges. These default to 64-bit floats in each entry.

maximelucas commented on xgi/linalg/laplacian_matrix.py:
```diff
@@ -183,13 +188,15 @@ def multiorder_laplacian(
     H : Hypergraph
         Hypergraph
+    weighted : bool, optional
+        whether or not to use hyperedge weightr, by default False (every edge weighted as 1).
```
```suggestion
        whether or not to use hyperedge weights, by default False (every edge weighted as 1).
```

maximelucas commented on xgi/linalg/laplacian_matrix.py:
```diff
@@ -221,16 +228,39 @@ def normalized_hypergraph_laplacian(H, sparse=True, index=False):
 
     if sparse:
-        Dinvsqrt = csr_array(diags(np.power(D, -0.5)))
+        Dv_invsqrt = csr_array(diags(np.power(Dv, -0.5)))
```
```suggestion
        Dv_invsqrt = diags_array(np.power(Dv, -0.5), format="csr")
```
Just saw [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.diags_array.html#scipy.sparse.diags_array) does it in one function and seems to be the new scipy go-to. Check that it works

maximelucas commented on xgi/linalg/laplacian_matrix.py:
```diff
@@ -221,16 +228,39 @@ def normalized_hypergraph_laplacian(H, sparse=True, index=False):
+    ) = incidence_matrix(H, sparse=sparse, index=True)
+
+    Dv = degree_matrix(H)
+    De = np.sum(incidence, axis=0)
```
```suggestion
    De = np.sum(incidence, axis=0)
    if weighted:
        weights = [H.edges[edge_idx].get("weight", 1) for edge_idx in H.edges]
    
```

maximelucas commented on xgi/linalg/laplacian_matrix.py:
```diff
@@ -221,16 +228,39 @@ def normalized_hypergraph_laplacian(H, sparse=True, index=False):
+        else:
+            W = identity(H.num_edges)
+
         eye = csr_array((H.num_nodes, H.num_nodes))
         eye.setdiag(1)
```
```suggestion
        eye = eye_array(H.num_nodes, H.num_nodes), format="csr")
```
Just realised this does the same in one line

maximelucas commented on xgi/linalg/laplacian_matrix.py:
```diff
@@ -221,16 +228,39 @@ def normalized_hypergraph_laplacian(H, sparse=True, index=False):
+        De_inv = np.diag(1 / De)
+
+        if weighted:
+            W = np.diag([H.edges[edge_idx].get("weight", 1) for edge_idx in H.edges])
```
```suggestion
            W = np.diag(weights)
```

maximelucas commented on xgi/linalg/laplacian_matrix.py:
```diff
@@ -221,16 +228,39 @@ def normalized_hypergraph_laplacian(H, sparse=True, index=False):
+        if weighted:
+            W = np.diag([H.edges[edge_idx].get("weight", 1) for edge_idx in H.edges])
+        else:
+            W = np.identity(H.num_edges)
```
```suggestion
            W = np.eye(H.num_edges)
```

maximelucas commented on xgi/linalg/laplacian_matrix.py:
```diff
@@ -221,16 +228,39 @@ def normalized_hypergraph_laplacian(H, sparse=True, index=False):
         eye = np.eye(H.num_nodes)
 
-    L = 0.5 * (eye - Dinvsqrt @ A @ Dinvsqrt)
+    L = eye - (
+        Dv_invsqrt @ incidence @ W @ De_inv @ np.transpose(incidence) @ Dv_invsqrt
+    )
```
```suggestion
    L = eye - Dv_invsqrt @ incidence @ W @ De_inv @ incidence.T @ Dv_invsqrt
```
Hopefully will stay on one line, .T should work on sparse and non-sparse

maximelucas commented on xgi/linalg/laplacian_matrix.py:
```diff
@@ -47,10 +47,15 @@
 
 import numpy as np
-from scipy.sparse import csr_array, diags
+from scipy.sparse import csr_array, diags, identity
```
update imports with new functions

maximelucas commented on tests/linalg/test_matrix.py:
```diff
@@ -605,25 +607,57 @@ def test_normalized_hypergraph_laplacian():
+    negative_evals = list(filter(lambda e: e < 0, evals))
+    assert (not negative_evals) or (np.allclose(negative_evals, 0))
 
+    L3, d = xgi.normalized_hypergraph_laplacian(H, index=True)
```
Sorry what I meant is I think we're not checking explicitly any of the sparse cases :) 
L2 is the sparse version of L3, maybe just make sure they're the same

maximelucas commented on xgi/linalg/laplacian_matrix.py:
```diff
@@ -107,7 +106,7 @@ def laplacian(H, order=1, sparse=False, rescale_per_node=False, index=False):
 
     if sparse:
-        K = csr_array(diags(degree_matrix(H, order=order)))
+        K = csr_array(diags_array(degree_matrix(H, order=order)))
```
```suggestion
        K = diags_array(degree_matrix(H, order=order), format="csr")
```

maximelucas commented:
Great, thanks a lot! Thanks for bearing with me :) 
I just left a last tiny comment (the `diags_array` is already a sparse matrix), but approved the PR already. 
Once you made the tiny change I'll squash and merge (I don't think you can, but if you can feel free to do it yourself).
Congrats on your first approved PR!

kaiser-dan commented:
> Great, thanks a lot! Thanks for bearing with me :) I just left a last tiny comment (the `diags_array` is already a sparse matrix), but approved the PR already. Once you made the tiny change I'll squash and merge (I don't think you can, but if you can feel free to do it yourself). Congrats on your first approved PR!

Thank you! It's been fun, I hope to contribute as much as I can moving forward as well.

nwlandry commented:
Nice work, @kaiser-dan! And thanks @maximelucas for reviewing!
